### PR TITLE
clientv3: initial lease

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @mitake @gyuho

### CODE REVIEW

File: clientv3/lease.go [hunk 1]
```diff
@@ -40,3 +49,230 @@ type Lease interface {
...
+		}
+
+		nerr := l.switchRemote(err)
+		if nerr != nil {
+			return nil, nerr
+		}
+	}
+}
+
+func (l *lessor) keepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
```
Comment (gyuho) [inline clientv3/lease.go]: Q. what do we mean by `Once` here?

Comment (xiang90) [inline clientv3/lease.go]: check the comment on the interface at line50

Comment (gyuho) [inline clientv3/lease.go]: Got it. Thanks.

On Fri, Jan 29, 2016, 17:37 Xiang Li <notifications@github.com> wrote:

> In clientv3/lease.go
> <https://github.com/coreos/etcd/pull/4353#discussion_r51337133>:
>
> > +
> > +func (l *lessor) KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
> > +	for {
> > +		resp, err := l.keepAliveOnce(ctx, id)
> > +		if err == nil {
> > +			return resp, err
> > +		}
> > +
> > +		nerr := l.switchRemote(err)
> > +		if nerr != nil {
> > +			return nil, nerr
> > +		}
> > +	}
> > +}
> > +
> > +func (l *lessor) keepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
>
> check the comment on the interface at line50
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4353/files#r51337133>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by gyuho** [comment]
lgtm

File: etcdctlv3/command/global.go [hunk 2]
```diff
@@ -37,13 +37,19 @@ func mustClient(cmd *cobra.Command) *clientv3.Client {
 	var cfgtls *transport.TLSInfo
 	tls := transport.TLSInfo{}
 	if tls.CertFile, err = cmd.Flags().GetString("cert"); err == nil {
-		cfgtls = &tls
+		if tls.CertFile != "" {
```
Comment (xiang90) [inline etcdctlv3/command/global.go]: /cc @heyitsanthony We should only set tls if the flag is not empty string.

I found there is another issue with current client library. We probably should let client library to create the initial conn, not accept one. Our client library should control how to create the grpc conn (tls or not, since it is in the config).

**Comment by xiang90** [comment]
@heyitsanthony I tested it. At least, it works and can recovers from connection failure... We need more extensive testing for sure.

File: etcdctlv3/command/global.go (hunk 2 above)
Comment (heyitsanthony) [inline etcdctlv3/command/global.go]: ok can you put this in another PR?

Comment (xiang90) [inline etcdctlv3/command/global.go]: sure. i will.

File: clientv3/lease.go [hunk 3]
```diff
@@ -26,17 +30,254 @@ type (
...
+	}
+
+	err = stream.Send(&pb.LeaseKeepAliveRequest{ID: int64(id)})
+	if err != nil {
+		stream.CloseSend()
+		return nil, err
+	}
+
+	resp, rerr := stream.Recv()
+	stream.CloseSend()
```
Comment (heyitsanthony) [inline clientv3/lease.go]: `CloseSend` isn't enough; stream's `ctx` must be cancelled or grpc will leak a goroutine

Comment (xiang90) [inline clientv3/lease.go]: good to know... I will cancel the ctx.

File: clientv3/lease.go [hunk 4]
```diff
@@ -26,17 +30,254 @@ type (
...
+			resp, err := stream.Recv()
+			if err != nil {
+				l.switchRemote(err)
+				break
+			}
+
+			l.mu.Lock()
+			lch, ok := l.keepAlives[lease.LeaseID(resp.ID)]
+			if !ok {
+				continue
```
Comment (heyitsanthony) [inline clientv3/lease.go]: unlock

Comment (xiang90) [inline clientv3/lease.go]: will fix. thanks!

File: clientv3/lease.go [hunk 5]
```diff
@@ -26,17 +30,254 @@ type (
...
+}
+
+func (l *lessor) getKeepAliveStream() pb.Lease_LeaseKeepAliveClient {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if l.stream != nil {
+		return l.stream
+	}
+
+	for {
```
Comment (heyitsanthony) [inline clientv3/lease.go]: any way to error out of this?

File: clientv3/lease.go [hunk 6]
```diff
@@ -26,17 +30,254 @@ type (
...
+		stream, err := l.remote.LeaseKeepAlive(context.TODO())
```
Comment (heyitsanthony) [inline clientv3/lease.go]: user-provided ctx here?

File: clientv3/lease.go (hunk 5 above)
Comment (xiang90) [inline clientv3/lease.go]: we probably should not error this out for the entire life-cycle of the client. the stream should always be here as long as the client exists. we should error this out when the client is closed. make sense? probably we should add a todo?

File: clientv3/lease.go (hunk 6 above)
Comment (xiang90) [inline clientv3/lease.go]: user should not control the internal long living keepAlive stream. we should give the client a lifecycle and when the user closes the client, we should cancel the stream.

File: clientv3/lease.go (hunk 5 above)
Comment (heyitsanthony) [inline clientv3/lease.go]: so long as there's some way to get out of this loop if the network is gone, maybe via `ctx` cancel via api consumer

Comment (xiang90) [inline clientv3/lease.go]: even if the network is gone, we should still keep retrying on this one so that when network recovers the long running keepalive send/recv loops can recover. we should give the user an option to stop the entire lease api life-cycle though, like Lease.Close() to stop all lease related resource?

File: clientv3/lease.go [hunk 7]
```diff
@@ -26,17 +30,254 @@ type (
...
+	resp, rerr := stream.Recv()
+	stream.CloseSend()
+	if rerr != nil {
+		return nil, rerr
+	}
+	return (*LeaseKeepAliveResponse)(resp), nil
+}
+
+func (l *lessor) recvKeepAliveLoop() {
+	for {
```
Comment (heyitsanthony) [inline clientv3/lease.go]: need way to break out of the loop for no goroutine leak

File: clientv3/lease.go [hunk 8]
```diff
@@ -26,17 +30,254 @@ type (
...
+				default:
+				}
+			}
+			l.mu.Unlock()
+		}
+	}
+}
+
+func (l *lessor) sendKeepAliveLoop() {
+	for {
```
Comment (heyitsanthony) [inline clientv3/lease.go]: also leaky

File: clientv3/lease.go (hunk 7 above)
Comment (xiang90) [inline clientv3/lease.go]: right. probably we should add lease.Close() or client.Close(). i am not sure which though.

File: clientv3/lease.go (hunk 5 above)
Comment (heyitsanthony) [inline clientv3/lease.go]: Yeah, I don't see it working correctly without a `Lease.Close()`

File: clientv3/lease.go (hunk 6 above)
Comment (heyitsanthony) [inline clientv3/lease.go]: ok

File: clientv3/lease.go (hunk 7 above)
Comment (heyitsanthony) [inline clientv3/lease.go]: There's already a `client.Close()` but it only works for the old grpc clients, it's broken under retries.

File: etcdctlv3/command/lease_command.go [hunk 9]
```diff
@@ -63,8 +62,9 @@ func leaseCreateCommandFunc(cmd *cobra.Command, args []string) {
 		ExitWithError(ExitBadArgs, fmt.Errorf("bad TTL (%v)", err))
 	}
 
-	req := &pb.LeaseCreateRequest{TTL: ttl}
-	resp, err := mustClient(cmd).Lease.LeaseCreate(context.Background(), req)
+	c := mustClient(cmd)
+	l := clientv3.NewLease(c)
```
Comment (heyitsanthony) [inline etcdctlv3/command/lease_command.go]: It would be easier to implement `client.Close()` if only clientv3 could call `newLease` and exposed the `Lease` as a field in `clientv3.Client` (same with the kv stuff too). Is there a use case for having multiple outstanding `Leases` per client?

File: clientv3/lease.go [hunk 10]
```diff
@@ -26,17 +30,254 @@ type (
...
+func (l *lessor) switchRemote(prevErr error) error {
+	newConn, err := l.c.retryConnection(l.conn, prevErr)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if l.stream != nil {
+		l.stream.CloseSend()
```
Comment (heyitsanthony) [inline clientv3/lease.go]: leaky if `l.stream` isn't already ctx-cancelled

File: clientv3/client.go [hunk 11]
```diff
@@ -169,6 +169,7 @@ func (c *Client) retryConnection(oldConn *grpc.ClientConn, err error) (*grpc.Cli
 		// conn has already been updated
 		return c.conn, nil
 	}
+	oldConn.Close()
```
Comment (heyitsanthony) [inline clientv3/client.go]: `oldConn` should already be closed/lost before calling retryConnection(), otherwise use of this function could devolve into rapidly flipping the connection on and off

Comment (xiang90) [inline clientv3/client.go]: > otherwise use of this function could devolve into rapidly flipping the connection on and off

Can you clarify this? I feel this can only happen if we wrongly call retryConnection frequently?

Comment (gyuho) [inline clientv3/client.go]: There's an experimental API that returns the status of a connection https://github.com/grpc/grpc-go/blob/master/clientconn.go#L239-L243. Might be useful?

Comment (xiang90) [inline clientv3/client.go]: @gyuho That API can return an error and it is implementation specific.

**Comment by xiang90** [comment]
@heyitsanthony PTAL. Thanks!

File: clientv3/lease.go [hunk 12]
```diff
@@ -26,17 +30,355 @@ type (
 	LeaseKeepAliveResponse pb.LeaseKeepAliveResponse
 )
 
+var (
```
Comment (heyitsanthony) [inline clientv3/lease.go]: `const`?

File: clientv3/lease.go [hunk 13]
```diff
@@ -26,17 +30,355 @@ type (
...
 	// KeepAlive keeps the given lease alive forever.
-	KeepAlive(ctx context.Context, id lease.LeaseID) (chan<- *LeaseKeepAliveResponse, error)
+	KeepAlive(ctx context.Context, id lease.LeaseID) (<-chan *LeaseKeepAliveResponse, error)
 
 	// KeepAliveOnce renews the lease once. In most of the cases, Keepalive
 	// should be used instead of KeepAliveOnce.
 	KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error)
+
+	// Lease keeps internal routines and connections for effecient communication with etcd server.
+	// After using Lease, call Close() to relase all related resources.
```
Comment (heyitsanthony) [inline clientv3/lease.go]: relase -> release

File: clientv3/lease.go [hunk 14]
```diff
@@ -26,17 +30,355 @@ type (
...
 
 	// KeepAlive keeps the given lease alive forever.
-	KeepAlive(ctx context.Context, id lease.LeaseID) (chan<- *LeaseKeepAliveResponse, error)
+	KeepAlive(ctx context.Context, id lease.LeaseID) (<-chan *LeaseKeepAliveResponse, error)
 
 	// KeepAliveOnce renews the lease once. In most of the cases, Keepalive
 	// should be used instead of KeepAliveOnce.
 	KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error)
+
+	// Lease keeps internal routines and connections for effecient communication with etcd server.
```
Comment (heyitsanthony) [inline clientv3/lease.go]: effecient -> efficient

File: clientv3/lease.go [hunk 15]
```diff
@@ -26,17 +30,355 @@ type (
...
+	return l.remote
+}
+
+func (l *lessor) getKeepAliveStream() pb.Lease_LeaseKeepAliveClient {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.stream
+}
+
+func (l *lessor) switchRemote(prevErr error) error {
```
Comment (heyitsanthony) [inline clientv3/lease.go]: `switchStream`?

File: clientv3/lease.go [hunk 16]
```diff
@@ -26,17 +30,355 @@ type (
...
+	stream, err := l.getRemote().LeaseKeepAlive(sctx)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.sc = cancel
+	l.stream = stream
```
Comment (heyitsanthony) [inline clientv3/lease.go]: this may leak a stream if there are two concurrent `newStream` calls

File: clientv3/lease.go [hunk 17]
```diff
@@ -26,17 +30,355 @@ type (
...
+// its internal resource.
+func cancelWhenStop(cancel context.CancelFunc, stopc <-chan struct{}) chan<- struct{} {
+	done := make(chan struct{}, 1)
+
+	go func() {
+		select {
+		case <-stopc:
+			cancel()
+		case <-done:
+			cancel()
```
Comment (heyitsanthony) [inline clientv3/lease.go]: just cancel() at the end of the goroutine instead of the cases?

File: clientv3/lease.go [hunk 18]
```diff
@@ -26,17 +30,354 @@ type (
...
+
+type lessor struct {
+	c *Client
+
+	mu     sync.Mutex       // guards all fields
+	conn   *grpc.ClientConn // conn in-use
+	remote pb.LeaseClient
+
+	stream pb.Lease_LeaseKeepAliveClient
+	sc     context.CancelFunc
```
Comment (heyitsanthony) [inline clientv3/lease.go]: `streamCancel` to be consistent with `stopCancel` naming

**Comment by xiang90** [comment]
@heyitsanthony All fixed. PTAL.

File: clientv3/lease.go (hunk 18 above)
Comment (xiang90) [inline clientv3/lease.go]: sure.

**Comment by heyitsanthony** [comment]
lgtm
